File: src/f/master/trig3dh.f
C
C *** LAST REVISED ON  4-JAN-1994 10:28:42.41
C *** SOURCE FILE: [LONGLIB93.SOURCES.FORTRAN.MASTER]TRIG3DH.FOR
C
	SUBROUTINE TRIG3DH(X,Y,Z,N,PSI,PHI,THETA,
     $		DV,XH,YH,ZH,IFLAG,IAXIS,
     $		XT,NXT,TX,TSX,PGX,FSX,
     $		YT,NYT,TY,TSY,PGY,FSY,
     $		ZT,NZT,TZ,TSZ,PGZ,FSZ,
     $		XAST,XAEN,YAST,YAEN,ZAST,ZAEN,ICOL)
C
C	CREATED BY D. LONG     FEB, 1986	AT JPL
C
C	ROUTINE TO PLOT DATA SPECIFIED AS (X,Y,Z) TRIPLES AS 3-D SURFACE
C	BY TRIANGULATING AND PLOTTING WITH HIDDEN LINE REMOVAL
C	USING THE INIT3DH ROUTINES.
C
C	X,Y,Z	 (R) ARRAYS CONTAINING DATA POINTS
C	N	 (I) NUMBER OF DATA POINTS
C	PSI,PHI,THETA (R) ANGLES (IN DEGREES) OF INIT3DH
C	DV	 (R) PERSPECTIVE FACTOR OF INIT3DH (9999=NO PERSPECTIVE)
C		    DV < 0 DO NOT INITIALIZE INIT3DH
C	XH,YH,ZH (R) LENGTH OF EACH AXIS (INCHES)
C	IFLAG	(I) PLOT FLAG
C		>1000	REPEAT PLOTTING BUT DO NOT INTIALIZE PLOTTER
C		< 0	REPEAT PLOTTING (DON'T CLOSE PLOTTER OUTPUT)
C		= 0	CLOSE PLOTTER OUTPUT--DO NOT PRODUCE PLOT
C		> 0	SINGLE PLOT OUTPUT ONLY (PLOTTER OUTPUT CLOSED)
C 	(ONE'S DIGIT) = 2 USE RAM TEK COLOR CONTROL ARRAY
C		      = 1 DO NOT USE RAM TEK COLOR ARRAY
C	(TEN'S DIGIT) = 0 NO HIDDEN LINE REMOVAL FOR AXIS
C		      = 1 HIDDEN LINE REMOVAL OF AXIS
C		      = 2 NO HIDDEN LINES AT ALL (SKETCH USED)
C		      = 3 NO HIDDEN LINES AT ALL (SKETCH NOT USED)
C 	(HUNDRED'S)   = 0 ASK WHICH SCREEN DEVICE TO USE
C		      <>0 SCREEN DEVICE CODE NUMBER
C 	IAXIS	(I) AXIS OPTION FLAG
C			< 0 DO NOT PLOT AXES
C 	(ONE'S DIGIT)	= 1 PLOT Z AXIS USING COMPUTED SCALE
C			= 2 PLOT Z AXIS USING COMPUTED, SMOOTHED SCALE
C			= 3 PLOT Z AXIS USING INPUT SCALE--VARIABLES ACCESSED
C			= 4 PLOT Z AXIS USING INPUT, SMOOTHED SCALE--VARIABLES ACCESSED
C 	(TEN'S DIGIT)	= 1 PLOT Y AXIS USING COMPUTED SCALE
C			= 2 PLOT Y AXIS USING COMPUTED, SMOOTHED SCALE
C			= 3 PLOT Y AXIS USING INPUT SCALE--VARIABLES ACCESSED
C			= 4 PLOT Y AXIS USING INPUT, SMOOTHED SCALE--VARIABLES ACCESSED
C 	(HUNDRED'S)	= 1 PLOT X AXIS USING COMPUTED SCALE
C			= 2 PLOT X AXIS USING COMPUTED, SMOOTHED SCALE
C			= 3 PLOT X AXIS USING INPUT SCALE--VARIABLES ACCESSED
C			= 4 PLOT X AXIS USING INPUT, SMOOTHED SCALE--VARIABLES ACCESSED
C 	XT,YT,ZT	(B) STRINGS FOR AXIS TITLES
C 	NXT,NYT,NZT	(I) LENGTH OF AXIS TITLES
C			     IF ZERO THEN AXIS NOT PLOTTED
C FOLLOWING ONLY ACCESSED IF IAXIS > 0
C	TX,TY,TZ	(R) NUMBER OF TICK MARKS (SEE AXIS3DH)
C	TSX,TSY,TSZ	(R) SIZE OF TITLE AND NUMBERS OF AXIS
C			  IF LESS THAN ZERO DO NOT AUTO-SCALE BY (x10^POWER)
C	PGX,PGY,PGZ	(R) ROTATION OF TEXT AND TICKS AROUND AXIS (DEG)
C	FSX,FSY,FSZ	(R) NUMBER FORMAT OF AXISES (SEE AXIS3DH)
C 	XAST,YAST,ZAST	(R) AXIS START VALUES
C 	XAEN,YAEN,ZAEN	(R) AXIS END VALUES
C	ICOL		(I) COLOR INDEX TABLE (ACCESSED IF 1'S DIGIT OF IFLAG=2
C				ICOL(1) AXIS COLOR
C				ICOL(2) SURFACE COLOR
C			    NOTE: WHEN BOTH AXIS AND SURFACE USE SKETCH
C			    BOTH ARE PLOTTED USING SURFACE COLOR
C
	DIMENSION X(1),Y(1),Z(1),PAS(2),ICOL(2)
	LOGICAL REPEAT,HIDDEN
	SAVE REPEAT
	CHARACTER XT,YT,ZT
C
	DIMENSION XX(4),YY(4),ZZ(4)
	DIMENSION ITT(2000,3),IP(1000),IB(1000)
C
	PARAMETER (IWRKSZE=100000)
	COMMON /GO/ISIZE,WORK(IWRKSZE)
	SAVE /GO/
	DATA NZZ/2000/,NW/1000/
	DATA REPEAT/.FALSE./
C
	IF (ISIZE.LT.IWRKSZE) ISIZE=IWRKSZE	! INCREASE WORKSPACE IF NEEDED
C
	IF (N.GT.NW) THEN
		CALL CTERM(1)
		WRITE (*,3010)
3010		FORMAT(' *** WARNING: INSUFFICIENT WORKING SPACE')
		CALL CTERM(-1)
	ENDIF
C
	IF (IFLAG.EQ.0) THEN
		IF (REPEAT) CALL PLOTND	!CLOSE PLOTTER
		REPEAT=.FALSE.
		RETURN
	ENDIF
C
	XLEN=ABS(XH)
	YLEN=ABS(YH)
	ZLEN=ABS(ZH)
	IZAXIS=MOD(IABS(IAXIS),10)
	IYAXIS=MOD(IABS(IAXIS),100)/10
	IXAXIS=MOD(IABS(IAXIS),1000)/100
	IF (IXAXIS.GT.2) THEN
		XMAX=XAEN
		XMIN=XAST
	ELSE
		XMAX=X(1)
		XMIN=X(1)
		DO 16 I=1,N
			XMAX=AMAX1(XMAX,X(I))
			XMIN=AMIN1(XMIN,X(I))
16		CONTINUE
	ENDIF
	IF (IXAXIS.EQ.2.OR.IXAXIS.EQ.4) THEN
		PAS(1)=XMAX
		PAS(2)=XMIN
		CALL SCALE(PAS,XLEN,2,1,1,XMIN,DAA)
		XMAX=YLEN*DAA+XMIN
	ENDIF
C
	IF (IYAXIS.GT.2) THEN
		YMAX=YAEN
		YMIN=YAST
	ELSE
		YMAX=Y(1)
		YMIN=Y(1)
		DO 17 I=1,N
			YMAX=AMAX1(YMAX,Y(I))
			YMIN=AMIN1(YMIN,Y(I))
17		CONTINUE
	ENDIF
	IF (IYAXIS.EQ.2.OR.IYAXIS.EQ.4) THEN
		PAS(1)=YMAX
		PAS(2)=YMIN
		CALL SCALE(PAS,YLEN,2,1,1,YMIN,DAA)
		YMAX=YLEN*DAA+YMIN
	ENDIF
C
	IF (IZAXIS.GT.2) THEN
		ZMAX=ZAEN
		ZMIN=ZAST
	ELSE
		ZMAX=Z(1)
		ZMIN=Z(1)
		DO 18 I=1,N
			ZMAX=AMAX1(ZMAX,Z(I))
			ZMIN=AMIN1(ZMIN,Z(I))
18		CONTINUE
	ENDIF
	IF (IZAXIS.EQ.2.OR.IZAXIS.EQ.4) THEN
		PAS(1)=ZMAX
		PAS(2)=ZMIN
		CALL SCALE(PAS,ZLEN,2,1,1,ZMIN,DAA)
		ZMAX=ZLEN*DAA+ZMIN
	ENDIF
C
C	INTIALIZE PLOT PACKAGE
C
	JF=IABS(IFLAG)
	IF (.NOT.REPEAT.AND.JF.LT.10000) THEN
		ILU=MOD(JF,10000)/100
		CALL FRAME(3,ILU,1.5,0.65,1.)		!INTIALIZE
	ELSE
		CALL CTERM(-1)				!PUT TERMINAL IN PLOT MODE
	ENDIF
C
	HIDDEN=.TRUE.
	IOPT=MOD(JF,100)/10
	IF (IOPT.EQ.3) THEN
		IOPT=0
		HIDDEN=.FALSE.
	ENDIF
	JF=MOD(JF,10)
C
C	INITIALIZE 3DH PACKAGE
C
      	MNE=3			! MAX NUMBER OF EDGES
      	MNH=0			! NO HOLES
	ICORE=(25+5*MNE+4*MNH)*N*N  ! dimension some space
	IF (IAXIS.NE.0) ICORE=ICORE+10000	! ESTIMATED AXISES SPACE
	IF (ICORE.GT.ISIZE) THEN
		CALL CTERM(1)
		WRITE (*,3011)
3011		FORMAT('*** WARNING: COMMON BLOCK /GO/ MAY BE TOO SMALL')
		CALL CTERM(-1)
	ENDIF
	SF=1.0			! SCALE FACTOR
	IF (IOPT.EQ.2) SF=-1.0	! NO HIDDEN LINE REMOVAL
	IF (DV.GT.0.) CALL INIT3DH(0.,0.,0.,PSI,PHI,THETA,SF,DV,MNE,MNH)
C
	IF (IAXIS.NE.0) THEN		!PLOT AXIS LABELS
		IF (JF.EQ.2) CALL PLOT(FLOAT(ICOL(1)),0.,0) !RAM TEK COLOR
		IF (NXT.GT.0) THEN 	!PLOT X AXIS
			CALL AXIS3DH(0.,0.,0.,0.,0.,PGX,XT,
     $				NXT,XLEN,XMIN,XMAX,
     $				TX,TSX,FSX,IOPT)
		ENDIF
		IF (NYT.GT.0) THEN 	!PLOT Y AXIS
			CALL AXIS3DH(0.,0.,0.,0.,90.,PGY,YT,
     $				NYT,YLEN,YMIN,YMAX,
     $				TY,TSY,FSY,IOPT)
		ENDIF
		IF (NZT.GT.0) THEN 	!PLOT Z AXIS
			CALL AXIS3DH(0.,0.,0.,90.,0.,PGZ,ZT,
     $				NZT,-ZLEN,ZMIN,ZMAX,
     $				TZ,TSZ,FSZ,IOPT)
		ENDIF
	ENDIF
	IF (JF.EQ.2) CALL PLOT(FLOAT(ICOL(2)),0.,0) !RAM TEK COLOR
C
C	BEGIN SURFACE PLOTTING
C
	DX=1.0
	DY=1.0
	DZ=1.0
	IF (XMAX-XMIN.NE.0.0) DX=XLEN/(XMAX-XMIN)
	IF (YMAX-YMIN.NE.0.0) DY=YLEN/(YMAX-YMIN)
	IF (ZMAX-ZMIN.NE.0.0) DZ=ZLEN/(ZMAX-ZMIN)
C
	CALL TRIANGC(X,Y,-N,ITT,NZZ,M,IP,IB,NW,L,NZZ,IE,IBE,ITE)
	DO 50 I=1,M
		DO 15 J=1,3
			K1=ITT(I,J)
			XX(J)=(X(K1)-XMIN)*DX
			YY(J)=(Y(K1)-YMIN)*DY
			ZZ(J)=(Z(K1)-ZMIN)*DY
15		CONTINUE
		XX(4)=XX(1)
		YY(4)=YY(1)
		ZZ(4)=ZZ(1)
		NC=4
		IERR=0
		IF (I.EQ.M) IERR=1
		IF (HIDDEN) THEN
			CALL SKETCH(XX,YY,ZZ,NC,IERR)
		ELSE
			IERR=3
			DO 25 JF=1,NC
			  CALL PLT3DH(XX(JF),YY(JF),ZZ(JF),IERR)
			  IERR=2
25			CONTINUE
		ENDIF
		IF (IERR.LT.0) THEN
			CALL CTERM(1)
			WRITE(*,3012) IERR
3012			FORMAT(' *** WARNING: SKETCH ERROR HAS OCCURED',I5)
			CALL CTERM(-1)
		ENDIF
 50	CONTINUE
	IF (.NOT.HIDDEN) CALL PLT3DH(XX(JF),YY(JF),ZZ(JF),3)	!PEN UP
C
C	FINISH ROUTINE
C
	CALL PLOT(0.,0.,3)			!PEN UP
	IF (IFLAG.GT.0) THEN
		CALL CTERM(2)			!CLEAR TERMINAL
		CALL PLOTND			!CLOSE PLOT OUTPUT
		REPEAT=.FALSE.
	ENDIF
	IF (IFLAG.LT.0) THEN
		REPEAT=.TRUE.
		CALL CTERM(1)			!PUT TERMINAL IN TEXT MODE
	ENDIF
	RETURN
	END

File: src/c/master/trig3dh.c

/* *** LAST REVISED ON  7-JAN-1994 12:56:16.27 */
/* *** SOURCE FILE: [LONGLIB93.SOURCES.C.MASTER]TRIG3DH.C */

#include "my.h"

/* Common Block Declarations */

extern struct {
    integer isize;
    real work[100000];
} go_;

#define go_1 go_

/* Table of constant values */

static integer c__1 = 1;
static integer c_n1 = -1;
static integer c__2 = 2;
static integer c__3 = 3;
static real c_b19 = (float)1.5;
static real c_b20 = (float).65;
static real c_b21 = (float)1.;
static real c_b27 = (float)0.;
static integer c__0 = 0;
static real c_b41 = (float)90.;

/* Subroutine */ int trig3dh_(x, y, z, n, psi, phi, theta, dv, xh, yh, zh, 
	iflag, iaxis, xt, nxt, tx, tsx, pgx, fsx, yt, nyt, ty, tsy, pgy, fsy, 
	zt, nzt, tz, tsz, pgz, fsz, xast, xaen, yast, yaen, zast, zaen, icol)
real *x, *y, *z;
integer *n;
real *psi, *phi, *theta, *dv, *xh, *yh, *zh;
integer *iflag, *iaxis;
char *xt;
integer *nxt;
real *tx, *tsx, *pgx, *fsx;
char *yt;
integer *nyt;
real *ty, *tsy, *pgy, *fsy;
char *zt;
integer *nzt;
real *tz, *tsz, *pgz, *fsz, *xast, *xaen, *yast, *yaen, *zast, *zaen;
integer *icol;
{
    /* Initialized data */

    static integer nzz = 2000;
    static integer nw = 1000;
    static logical repeat = FALSE_;

    /* System generated locals */
    integer i__1, i__2;
    real r__1, r__2;

    /* Local variables */
    static integer ierr;
    static real xlen, ylen, zlen, xmin, ymin, xmax, ymax, zmax, zmin;
    static integer iopt;
    extern /* Subroutine */ int plot_();
    static integer i, j, l, m;
    extern /* Subroutine */ int scale_(), frame_();
    static integer icore;
    extern /* Subroutine */ int cterm_();
    static integer k1;
    extern /* Subroutine */ int plt3dh_();
    static integer ib[1000], ie, jf, nc, ip[1000];
    static logical hidden;
    static real sf, dx, dy;
    static real xx[4], yy[4], zz[4];
    extern /* Subroutine */ int sketch_(), plotnd_();
    static integer ixaxis, iyaxis, izaxis;
    extern /* Subroutine */ int init3dh_(), axis3dh_();
    static real daa;
    static integer ibe, mne, mnh;
    static real pas[2];
    static integer ite, ilu, itt[6000]	/* was [2000][3] */;
    extern /* Subroutine */ int triangc_();

/* 	CREATED BY D. LONG     FEB, 1986	AT JPL */

/* 	ROUTINE TO PLOT DATA SPECIFIED AS (X,Y,Z) TRIPLES AS 3-D SURFACE */
/* 	BY TRIANGULATING AND PLOTTING WITH HIDDEN LINE REMOVAL */
/* 	USING THE INIT3DH ROUTINES. */

/* 	X,Y,Z	 (R) ARRAYS CONTAINING DATA POINTS */
/* 	N	 (I) NUMBER OF DATA POINTS */
/* 	PSI,PHI,THETA (R) ANGLES (IN DEGREES) OF INIT3DH */
/* 	DV	 (R) PERSPECTIVE FACTOR OF INIT3DH (9999=NO PERSPECTIVE) */
/* 		    DV < 0 DO NOT INITIALIZE INIT3DH */
/* 	XH,YH,ZH (R) LENGTH OF EACH AXIS (INCHES) */
/* 	IFLAG	(I) PLOT FLAG */
/* 		>1000	REPEAT PLOTTING BUT DO NOT INTIALIZE PLOTTER */
/* 		< 0	REPEAT PLOTTING (DON'T CLOSE PLOTTER OUTPUT) */
/* 		= 0	CLOSE PLOTTER OUTPUT--DO NOT PRODUCE PLOT */
/* 		> 0	SINGLE PLOT OUTPUT ONLY (PLOTTER OUTPUT CLOSED) */
/* 	(ONE'S DIGIT) = 2 USE RAM TEK COLOR CONTROL ARRAY */
/* 		      = 1 DO NOT USE RAM TEK COLOR ARRAY */
/* 	(TEN'S DIGIT) = 0 NO HIDDEN LINE REMOVAL FOR AXIS */
/* 		      = 1 HIDDEN LINE REMOVAL OF AXIS */
/* 		      = 2 NO HIDDEN LINES AT ALL (SKETCH USED) */
/* 		      = 3 NO HIDDEN LINES AT ALL (SKETCH NOT USED) */
/* 	(HUNDRED'S)   = 0 ASK WHICH SCREEN DEVICE TO USE */
/* 		      <>0 SCREEN DEVICE CODE NUMBER */
/* 	IAXIS	(I) AXIS OPTION FLAG */
/* 			< 0 DO NOT PLOT AXES */
/* 	(ONE'S DIGIT)	= 1 PLOT Z AXIS USING COMPUTED SCALE */
/* 			= 2 PLOT Z AXIS USING COMPUTED, SMOOTHED SCALE */
/* 			= 3 PLOT Z AXIS USING INPUT SCALE--VARIABLES ACCESSED */
/* 			= 4 PLOT Z AXIS USING INPUT, SMOOTHED SCALE--VARIABLES ACCESSED */
/* 	(TEN'S DIGIT)	= 1 PLOT Y AXIS USING COMPUTED SCALE */
/* 			= 2 PLOT Y AXIS USING COMPUTED, SMOOTHED SCALE */
/* 			= 3 PLOT Y AXIS USING INPUT SCALE--VARIABLES ACCESSED */
/* 			= 4 PLOT Y AXIS USING INPUT, SMOOTHED SCALE--VARIABLES ACCESSED */
/* 	(HUNDRED'S)	= 1 PLOT X AXIS USING COMPUTED SCALE */
/* 			= 2 PLOT X AXIS USING COMPUTED, SMOOTHED SCALE */
/* 			= 3 PLOT X AXIS USING INPUT SCALE--VARIABLES ACCESSED */
/* 			= 4 PLOT X AXIS USING INPUT, SMOOTHED SCALE--VARIABLES ACCESSED */
/* 	XT,YT,ZT	(B) STRINGS FOR AXIS TITLES */
/* 	NXT,NYT,NZT	(I) LENGTH OF AXIS TITLES */
/* 			     IF ZERO THEN AXIS NOT PLOTTED */
/* FOLLOWING ONLY ACCESSED IF IAXIS > 0 */
/* 	TX,TY,TZ	(R) NUMBER OF TICK MARKS (SEE AXIS3DH) */
/* 	TSX,TSY,TSZ	(R) SIZE OF TITLE AND NUMBERS OF AXIS */
/* 			  IF LESS THAN ZERO DO NOT AUTO-SCALE BY (x10^POWER) */
/* 	PGX,PGY,PGZ	(R) ROTATION OF TEXT AND TICKS AROUND AXIS (DEG) */
/* 	FSX,FSY,FSZ	(R) NUMBER FORMAT OF AXISES (SEE AXIS3DH) */
/* 	XAST,YAST,ZAST	(R) AXIS START VALUES */
/* 	XAEN,YAEN,ZAEN	(R) AXIS END VALUES */
/* 	ICOL		(I) COLOR INDEX TABLE (ACCESSED IF 1'S DIGIT OF IFLAG=2 */
/* 				ICOL(1) AXIS COLOR */
/* 				ICOL(2) SURFACE COLOR */
/* 			    NOTE: WHEN BOTH AXIS AND SURFACE USE SKETCH */
/* 			    BOTH ARE PLOTTED USING SURFACE COLOR */

    /* Parameter adjustments */
    --icol;
    --z;
    --y;
    --x;

    /* Function Body */

    if (go_1.isize < 100000) {
	go_1.isize = 100000;
    }

/* INCREASE WORKSPACE IF NEEDED */
    if (*n > nw) {
	cterm_(&c__1);
	printf("\n*** TRIG3DH: WARNING, STORAGE /GO/ TOO SMALL ***\n");
	cterm_(&c_n1);
    }

    if (*iflag == 0) {
	if (repeat) {
	    plotnd_();
	}
/* CLOSE PLOTTER */
	repeat = FALSE_;
	return 0;
    }

    xlen = dabs(*xh);
    ylen = dabs(*yh);
    zlen = dabs(*zh);
    izaxis = abs(*iaxis) % 10;
    iyaxis = abs(*iaxis) % 100 / 10;
    ixaxis = abs(*iaxis) % 1000 / 100;
    if (ixaxis > 2) {
	xmax = *xaen;
	xmin = *xast;
    } else {
	xmax = x[1];
	xmin = x[1];
	i__1 = *n;
	for (i = 1; i <= i__1; ++i) {
/* Computing MAX */
	    r__1 = xmax, r__2 = x[i];
	    xmax = dmax(r__1,r__2);
/* Computing MIN */
	    r__1 = xmin, r__2 = x[i];
	    xmin = dmin(r__1,r__2);
/* L16: */
	}
    }
    if (ixaxis == 2 || ixaxis == 4) {
	pas[0] = xmax;
	pas[1] = xmin;
	scale_(pas, &xlen, &c__2, &c__1, &c__1, &xmin, &daa);
	xmax = ylen * daa + xmin;
    }

    if (iyaxis > 2) {
	ymax = *yaen;
	ymin = *yast;
    } else {
	ymax = y[1];
	ymin = y[1];
	i__1 = *n;
	for (i = 1; i <= i__1; ++i) {
/* Computing MAX */
	    r__1 = ymax, r__2 = y[i];
	    ymax = dmax(r__1,r__2);
/* Computing MIN */
	    r__1 = ymin, r__2 = y[i];
	    ymin = dmin(r__1,r__2);
/* L17: */
	}
    }
    if (iyaxis == 2 || iyaxis == 4) {
	pas[0] = ymax;
	pas[1] = ymin;
	scale_(pas, &ylen, &c__2, &c__1, &c__1, &ymin, &daa);
	ymax = ylen * daa + ymin;
    }

    if (izaxis > 2) {
	zmax = *zaen;
	zmin = *zast;
    } else {
	zmax = z[1];
	zmin = z[1];
	i__1 = *n;
	for (i = 1; i <= i__1; ++i) {
/* Computing MAX */
	    r__1 = zmax, r__2 = z[i];
	    zmax = dmax(r__1,r__2);
/* Computing MIN */
	    r__1 = zmin, r__2 = z[i];
	    zmin = dmin(r__1,r__2);
/* L18: */
	}
    }
    if (izaxis == 2 || izaxis == 4) {
	pas[0] = zmax;
	pas[1] = zmin;
	scale_(pas, &zlen, &c__2, &c__1, &c__1, &zmin, &daa);
	zmax = zlen * daa + zmin;
    }

/* 	INTIALIZE PLOT PACKAGE */

    jf = abs(*iflag);
    if (! repeat && jf < 10000) {
	ilu = jf % 10000 / 100;
	frame_(&c__3, &ilu, &c_b19, &c_b20, &c_b21);
/* INTIALIZE */
    } else {
	cterm_(&c_n1);
/* PUT TERMINAL IN PLOT MODE */
    }

    hidden = TRUE_;
    iopt = jf % 100 / 10;
    if (iopt == 3) {
	iopt = 0;
	hidden = FALSE_;
    }
    jf %= 10;

/* 	INITIALIZE 3DH PACKAGE */

    mne = 3;
/* MAX NUMBER OF EDGES */
    mnh = 0;
/* NO HOLES */
    icore = (mne * 5 + 25 + (mnh << 2)) * *n * *n;
    if (*iaxis != 0) {
	icore += 10000;
    }
/* ESTIMATED AXISES SPACE */
    if (icore > go_1.isize) {
	cterm_(&c__1);
	printf("\n*** TRIG3DH: WARNING, STORAGE /GO/ TOO SMALL ***\n");
	cterm_(&c_n1);
    }
    sf = (float)1.;
/* SCALE FACTOR */
    if (iopt == 2) {
	sf = (float)-1.;
    }
/* NO HIDDEN LINE REMOVAL */
    if (*dv > (float)0.) {
	init3dh_(&c_b27, &c_b27, &c_b27, psi, phi, theta, &sf, dv, &mne, &mnh)
		;
    }

    if (*iaxis != 0) {
/* PLOT AXIS LABELS */
	if (jf == 2) {
	    r__1 = (real) icol[1];
	    plot_(&r__1, &c_b27, &c__0);
	}
/* RAM TEK COLOR */
	if (*nxt > 0) {
/* PLOT X AXIS */
	    axis3dh_(&c_b27, &c_b27, &c_b27, &c_b27, &c_b27, pgx, xt, nxt, &
		    xlen, &xmin, &xmax, tx, tsx, fsx, &iopt, 1L);
	}
	if (*nyt > 0) {
/* PLOT Y AXIS */
	    axis3dh_(&c_b27, &c_b27, &c_b27, &c_b27, &c_b41, pgy, yt, nyt, &
		    ylen, &ymin, &ymax, ty, tsy, fsy, &iopt, 1L);
	}
	if (*nzt > 0) {
/* PLOT Z AXIS */
	    r__1 = -(doublereal)zlen;
	    axis3dh_(&c_b27, &c_b27, &c_b27, &c_b41, &c_b27, pgz, zt, nzt, &
		    r__1, &zmin, &zmax, tz, tsz, fsz, &iopt, 1L);
	}
    }
    if (jf == 2) {
	r__1 = (real) icol[2];
	plot_(&r__1, &c_b27, &c__0);
    }

/* 	BEGIN SURFACE PLOTTING */

/* RAM TEK COLOR */
    dx = (float)1.;
    dy = (float)1.;
    if (xmax - xmin != (float)0.) {
	dx = xlen / (xmax - xmin);
    }
    if (ymax - ymin != (float)0.) {
	dy = ylen / (ymax - ymin);
    }

    i__1 = -(*n);
    triangc_(&x[1], &y[1], &i__1, itt, &nzz, &m, ip, ib, &nw, &l, &nzz, &ie, &
	    ibe, &ite);
    i__1 = m;
    for (i = 1; i <= i__1; ++i) {
	for (j = 1; j <= 3; ++j) {
	    k1 = itt[i + j * 2000 - 2001];
	    xx[j - 1] = (x[k1] - xmin) * dx;
	    yy[j - 1] = (y[k1] - ymin) * dy;
	    zz[j - 1] = (z[k1] - zmin) * dy;
/* L15: */
	}
	xx[3] = xx[0];
	yy[3] = yy[0];
	zz[3] = zz[0];
	nc = 4;
	ierr = 0;
	if (i == m) {
	    ierr = 1;
	}
	if (hidden) {
	    sketch_(xx, yy, zz, &nc, &ierr);
	} else {
	    ierr = 3;
	    i__2 = nc;
	    for (jf = 1; jf <= i__2; ++jf) {
		plt3dh_(&xx[jf - 1], &yy[jf - 1], &zz[jf - 1], &ierr);
		ierr = 2;
/* L25: */
	    }
	}
	if (ierr < 0) {
	    cterm_(&c__1);
	    printf("\n*** TRIG3DH: SKETCH ERROR HAS OCCURRED ***\n");
	    cterm_(&c_n1);
	}
/* L50: */
    }
    if (! hidden) {
	plt3dh_(&xx[jf - 1], &yy[jf - 1], &zz[jf - 1], &c__3);
    }

/* 	FINISH ROUTINE */

/* PEN UP */
/* L999: */
    plot_(&c_b27, &c_b27, &c__3);
/* PEN UP */
    if (*iflag > 0) {
	cterm_(&c__2);
/* CLEAR TERMINAL */
	plotnd_();
/* CLOSE PLOT OUTPUT */
	repeat = FALSE_;
    }
    if (*iflag < 0) {
	repeat = TRUE_;
	cterm_(&c__1);
/* PUT TERMINAL IN TEXT MODE */
    }
    return 0;
} /* trig3dh_ */

